CONTRIBUTING.md

```diff
@@ -15,11 +15,36 @@ to discuss proposals and use pull requests to suggest changes. For more
 information see the [Cilium Development
 Guide](https://docs.cilium.io/en/latest/contributing/development/).
 
+## An important note about mutating cluster resources
+
+Cilium CLI must follow these general rules regarding mutating cluster resources:
+
+- Cilium CLI must not mutate resources that are not managed by the Cilium Helm
+  chart. Cilium CLI must use Cilium Helm values to configure Cilium installations.
+- The only exception to the rule above is the connectivity test command, which
+  may mutate resources within Kubernates namespaces with the name prefix specified
+  by --test-namespace flag. The connectivity test must not mutate resources that
+  affect network connectivity outside the test namespaces.
+- TODO: I'd like to deprecate the use of --include-unsafe-tests flag for adding
+  tests that mutate resources outside the test namespaces:
+  - These unsafe tests typically can't run in parallel.
+  - It sets a bad precedent for new contributors. They might see test cases in
+    the connectivity test command that mutate ressources outside the test
+    namespaces, and assume that the connectivity test is the right command to
+    add more unsafe tests. It only takes one careless review for these tests
+    that are not guarded by --include-unsafe-tests to slip in.
```

Comment:
What would be the threshold for how we structure the code that would make you feel more comfortable with protecting user clusters against mutable state?

I could imagine us building into the common libraries for `cilium connectivity test` that the k8s client returns an error if it tries to mutate anything outside the set of resources explicitly declared in the point above this one. We could do this with a simple wrapper around the current k8s client. Then it would be structurally difficult for someone to introduce unsafe logic into the connectivity tests unless they went and modified that common library as well. This would signal the danger more easily I think. We could even restrict the codeowners on that common library to ensure review is more tightly watched, while relaxing codeowners for general test improvements.

Then let's say there's another command like your proposed `cilium test`, that command would get unrestricted access to the k8s client. So it could do whatever it needs to.

Reply:
Can we give cilium-cli more restrictive permission when end-users run the connectivity test in their cluster. For example,  prohibit changes to kube-system namespace using k8s RBAC

Reply:
Interesting point. That's possible today, but it's up to the user to provide the restricted permissions to the CLI. The user could always provision and provide superadmin.

Reply:
^ That said, let's say we provisioned the RBAC in the testsuite with those permissions... it could provide a safeguard, because if the CLI attempted to create resources beyond that scope, it would fail with an RBAC error. Would have to specifically run a testsuite with such RBAC and without the `--include-unsafe-tests` flag.